extractor/main.go

```diff
@@ -0,0 +1,293 @@
+	if cfg.sourceDir == "" {
+		return errors.New("invalid source directory")
+	}
+
+	// Check the output dir is valid
+	if cfg.outputDir == "" {
+		return errors.New("invalid output directory")
+	}
+
+	// Find the files that need to be analyzed
+	sourceFiles, findErr := findFilePaths(cfg.sourceDir, cfg.fileType)
+	if findErr != nil {
+		return fmt.Errorf("unable to find file paths, %w", findErr)
+	}
+
+	// Concurrently process the source files
+	g, ctx := errgroup.WithContext(ctx)
+
+	for _, sourceFile := range sourceFiles {
+		sourceFile := sourceFile
+
+		g.Go(func() error {
+			// Extract messages
+			msgs, processErr := extractAddMessages(sourceFile)
+			if processErr != nil {
+				return processErr
+			}
+
+			// Process messages
+			for _, msg := range msgs {
+				outputDir := filepath.Join(cfg.outputDir, strings.TrimLeft(msg.Package.Path, "gno.land/"))
+
+				// Write dir before writing files
+				if dirWriteErr := os.MkdirAll(outputDir, os.ModePerm); dirWriteErr != nil {
+					return fmt.Errorf("unable to write dir, %w", dirWriteErr)
+				}
+
+				// Write the package source code
+				if writeErr := writePackageFiles(msg, outputDir); writeErr != nil {
+					return writeErr
+				}
+
+				// Write the package metadata
+				if writeErr := writePackageMetadata(metadataFromMsg(msg), outputDir); writeErr != nil {
+					return writeErr
+				}
+			}
+
+			return nil
+		})
+	}
+
+	return g.Wait()
+}
+
+// writePackageFiles writes all files from a single package to the output directory
+func writePackageFiles(msg vm.MsgAddPackage, outputDir string) error {
+	for _, file := range msg.Package.Files {
+		// Get the output path
+		writePath := filepath.Join(outputDir, file.Name)
+
+		if writeErr := os.WriteFile(writePath, []byte(file.Body), 0644); writeErr != nil {
+			return fmt.Errorf("unable to write file %s, %w", file.Name, writeErr)
+		}
+	}
+
+	return nil
+}
+
+// writePackageMetadata writes the package metadata to the output directory
+func writePackageMetadata(metadata Metadata, outputDir string) error {
+	// Get the output path
+	writePath := filepath.Join(outputDir, packageMetadataFile)
+
+	// Get the JSON metadata
+	metadataRaw, marshalErr := json.Marshal(metadata)
+	if marshalErr != nil {
+		return fmt.Errorf("unable to JSON marshal metadata, %w", marshalErr)
+	}
+
+	if writeErr := os.WriteFile(writePath, metadataRaw, 0644); writeErr != nil {
+		return fmt.Errorf("unable to write package metadata, %w", writeErr)
+	}
+
+	return nil
+}
+
+func extractAddMessages(filePath string) ([]vm.MsgAddPackage, error) {
+	file, openErr := os.Open(filePath)
+	if openErr != nil {
+		return nil, fmt.Errorf("unable to open file, %w", openErr)
+	}
+
+	cleanup := func() error {
+		if closeErr := file.Close(); closeErr != nil {
+			return fmt.Errorf("unable to gracefully close file, %w", closeErr)
+		}
+		return nil
+	}
+
+	reader := bufio.NewReader(file)
+
+	// Used to track what was parsed in the past
+	touchMap := make(map[string]vm.MsgAddPackage)
```

Comment:
Nit: I think you are not using the values, so you can reduce the memory footprint doing:
```suggestion
	touchMap := make(map[string]bool)
```